### ISSUE TITLE
Tracking Issue for precise pre-release update

### ISSUE THREAD

**Issue by ehuss** [rust-lang/cargo#13290]
### Summary

RFC: [#3493](https://github.com/rust-lang/rfcs/pull/3493)
Implementation: TODO
Documentation: TODO

Extends `cargo update --precise` to allow pre-release versions, even if the dependency declarations do not specify a pre-release.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

**Comment by linyihai** [comment]
Since https://github.com/rust-lang/cargo/pull/13296 had merged, i think i can lend a hand to continue.

Just in case I forgot, I was gonna @rustbot claim.

**Comment by epage** [comment]
I believe @eopb was planning on working on this.  I'd recommend coordinating with them

**Comment by eopb** [comment]
Yes, I'm still intending to work on this. If you'd like to help @linyihai, please reach out to me on zulip (under the same username as here).

I'm sure we can work together. The last thing I'd want to do is to slow things down by trying to hog this tracking issue.

**Comment by linyihai** [comment]
FYI，I recently tried to implement an MVP based on RFC content.

**Comment by weihanglo** [comment]
<https://github.com/rust-lang/cargo/pull/14013> shows the current implementation is a bit buggy, and there are upper bound semantic issues not yet resolved. See these test cases:

<https://github.com/rust-lang/cargo/blob/050f88c3316178a064619b79ef9f56f0583488cc/src/cargo/util/semver_ext.rs#L204-L233>

It is technically possible to fix them inside cargo. However to do that we'll need to duplicate the logic in `semver` crate. I am in the team of upstreaming things to `semver`.

**Comment by linyihai** [comment]
Yes, The `semver` crate only supports to compare prerelease with prerelease if I remember correct，
https://docs.rs/semver/latest/src/semver/eval.rs.html#14-21

```
 // If a version has a prerelease tag (for example, 1.2.3-alpha.3) then it
    // will only be allowed to satisfy req if at least one comparator with the
    // same major.minor.patch also has a prerelease tag.
    for cmp in &req.comparators {
        if pre_is_compatible(cmp, ver) {
            return true;
        }
    }
```

So we need to add the pre tag to the `version_req` in order to use the `semver` crate,  like RFC writes:

```
So before,

1.2.3  -> ^1.2.3 -> >=1.2.3, <2.0.0 (with implicit holes excluding pre-release versions)
would become

1.2.3  -> ^1.2.3 -> >=1.2.3, <2.0.0-0

```

**Comment by linyihai** [comment]
After some experiments on `semver` [PR,](https://github.com/dtolnay/semver/pull/321) I found something interesting, 

- the `semver` crate `matches` implementaion may be based on [node semver compatibility](https://github.com/npm/node-semver?tab=readme-ov-file#semver1----the-semantic-versioner-for-npm).  the semver `matches` follow the [limit ](https://github.com/npm/node-semver?tab=readme-ov-file#prerelease-tags) if the version has a prerelease tag then its MAJOR.MINOR.PATCH must be same with the VersionReq at least once,  and another issuse is it can't handle partial `VersionReq`
- By removing the [limit ](https://github.com/npm/node-semver?tab=readme-ov-file#prerelease-tags) and handle the partial `VersionReq`, the [first proposed semantic ](https://github.com/dtolnay/semver/pull/321#discussion_r1688962017) came out.  The prerelease tag padding logic has some controversial, like '<4.2', we chose pading it to '<4.2.0' not '<4.2.0-0' . 

By look into the `node semver compatibility` semantic, it offers [includePrerelease=true](https://github.com/npm/node-semver?tab=readme-ov-file#functions) to remove the limit, then I add the test and try to implement it in semver, see https://github.com/dtolnay/semver/pull/321#issuecomment-2249961396. The result is the `node semver compatibility` implement ation also covers the semantic proposed by https://github.com/rust-lang/rfcs/pull/3493.

**Comment by BatmanAoD** [comment]
Until this is implemented, is there any workaround to force Cargo to use a prerelease version of a crate that is required (using a non-prerelease string) by a dependency? I.e.:

```
foo
├── bar
│   └── tertiary v1.12.3
├── tertiary v1.13.0-dev.0
...
```

Here, is there any way to generate a valid `Cargo.lock` file without forking `bar`?

**Comment by weihanglo** [comment]
@BatmanAoD this is implemented in <https://github.com/rust-lang/cargo/pull/13626> and there is doc link in the PR description. This PR <https://github.com/rust-lang/cargo/pull/14305> also shows the current proposed semantic of matching pre-releases if you're interested.

Mind sharing your use case with minimal repro, if the above can't satisfied your need?

**Comment by BatmanAoD** [comment]
@weihanglo Sorry, I forgot you had asked for a reproduction, and I ran into this problem again. Here's a reproduction: https://github.com/BatmanAoD/cargo-prerelease-demo/tree/main

From the RFC, it looks like the `update` command I gave in the README _should_ work, but it does not.

**Comment by BatmanAoD** [comment]
@weihanglo Actually, with `-Zunstable-options`, it appears that the `update` command itself no longer exits with an error, but it doesn't actually modify `Cargo.lock`, and as far as I can tell, there's no way to actually build the workspace.

**Comment by ankurmittal** [comment]
Facing same issue. This prerelease and stable version conflict is in weird state. Cargo treats prerelease as incompatible (like multiple major versions), but does not allow you to have both pre-release and stable version (from transitive deps) even though it allows you to have multiple major versions

For example if my dep `bar` bring in `foo` at 0.8.0 and i try to update to 0.8.2-rc.9, it conflicts, but updating to `0.9` is allowed as long as public API for `bar` does not accept anything from `foo`. If cargo treats pre-release version as incompatible, it should atleast treat it as a separate major version.

**Comment by celinval** [comment]
I was playing with this feature a bit, and I ran into a surprising behavior, and I was wondering if that was by design. Basically, `cargo check` updates my lockfile if I select a release candidate to a version that has already been released.

Here is what I did:

1. Created an empty project, added hyper as a dependency with "*" and generated the lockfile.
```
cargo new --lib mylib
echo 'hyper = "*"' > mylib/Cargo.toml
cd mylib
cargo generate-lockfile
```

`Cargo.lock` correctly included v1.11.0 (latest version available when I ran this).

2. I forced the update using `--precise` which correctly **downgraded** hyper to the release candidate. Lockfile was correctly updated.

```
cargo +nightly update --precise 1.0.0-rc.1 hyper -Z unstable-options
```

3. I tried to check the changes:
```
cargo check
```

The lockfile was updated back to v1.11.0. So `cargo check` didn't actually respected the lockfile. If I pass `--locked` or `--frozen`, Cargo fails with the following error:

```
error: cannot update the lock file <path/to/Cargo.lock> because --locked was passed to prevent this
```

**Comment by epage** [comment]
Does the lockfile change if you `cargo check` with this feature enabled?

As for the feature itself, I worry if making exceptions for `--precise`  is a dead end in the curreat cli becauee it can't handle dependency trees. I see that `uv` has done some pre-release resolver work and wonder if any of it is relevant for a path forward for the use case.

**Comment by weihanglo** [comment]
I guess there is a bug:

https://github.com/rust-lang/cargo/blob/514c56dd7321eecbfdcf9b6479519cf4edfab906/src/ops/resolve.rs#L732-L736

The `dep` is directly from manifest, and cargo doesn't call `matches_prerelease` at all there. The `matches_ignoring_source` would match nothing.


```console
$ CARGO_LOG=cargo::ops::resolve=debug cargo check
   0.053556625s DEBUG resolve_with_registry:resolve_with_previous:register_patch_entries: cargo::ops::resolve: return=Ok({})
   0.053593791s DEBUG resolve_with_registry:resolve_with_previous:register_previous_locks: cargo::ops::resolve: poisoning registry `crates-io` because mylib v0.1.0 (/projects/mylib) looks like it changed hyper
   0.053635166s DEBUG resolve_with_registry:resolve_with_previous:register_previous_locks: cargo::ops::resolve: ignoring any lock pointing directly at bytes v1.12.1
   0.053638916s DEBUG resolve_with_registry:resolve_with_previous:register_previous_locks: 
...
   0.053691833s DEBUG resolve_with_registry:resolve_with_previous: cargo::ops::resolve: attempting to prefer bytes v1.12.1
   0.053694708s DEBUG resolve_with_registry:resolve_with_previous: cargo::ops::resolve: attempting to prefer fnv v1.0.7
...
    Updating crates.io index
     Locking 1 package to latest compatible version
    Updating hyper v1.0.0-rc.1 -> v1.11.0
```

Mind the "poisoning registry `crates-io` because mylib v0.1.0 (/projects/mylib) looks like it changed hyper"

**Comment by celinval** [comment]
Here is another weird one, which may have the same root cause, not sure. If the Cargo.lock version is a release candidate, the follow up `cargo update --precise <VERSION>` completely ignores the given version and updates the package version to the latest one.

```
$ cargo +nightly update --precise 1.0.0-rc.1 hyper -Z unstable-options                                                                                                                                                                                                                                                                   
    Updating crates.io index                                                                                                                                                                                                                                                                                                                                                        
      Adding ...
 Downgrading hyper v1.11.0 -> v1.0.0-rc.1                                                                                                                                                                                                                                                                                                                                           
      Adding ...
$ cargo +nightly update --precise 1.0.1 hyper                                                                                                                                                                                                                                                                                                     
    Updating crates.io index                                                                                                                                                                                                                                                                                                                                                        
    Removing ...                                                                                                                                                                                                                                                                                                                                                        
    Updating hyper v1.0.0-rc.1 -> v1.11.0                                                                                                                                                                                                                                                                                                                                           
    Removing ...
```

Note that the same behavior happens with or without `+nightly`.

> Does the lockfile change if you cargo check with this feature enabled?

@epage I don't think there's an unstable feature `precise-pre-release`. The option seems to be simply gated by `-Z unstable-options`.

**Comment by weihanglo** [comment]
The fix is not too hard. We basically just need to match prerelease in the below place (and perhaps some other lockfile reader places):

> I guess there is a bug:
> 
> https://github.com/rust-lang/cargo/blob/514c56dd7321eecbfdcf9b6479519cf4edfab906/src/ops/resolve.rs#L732-L736
> 
> ...
> 
> Mind the "poisoning registry `crates-io` because mylib v0.1.0 (/projects/mylib) looks like it changed hyper"

However if we did this, we silently allow this behavior on stable. That is, non prerelease requirement allows prerelease in lockfile.

We could let people opt-in this feature. However, this feature is now behind `-Zunstable-options` for only `cargo update`. We could switch back to a dedicated `-Zprerelease` unstable flag, so people can opt-in this behavior for commands like `cargo build` and other build commands. What do people think?
